core/input/keyboard.js

```diff
@@ -87,6 +89,37 @@ export default class Keyboard {
         const code = this._getKeyCode(e);
         let keysym = KeyboardUtil.getKeysym(e);
 
+        // If remote state of capslock is known, and it doesn't match the local led state of
+        // the keyboard, we send a capslock keypress first to bring it into sync.
+        // If we just pressed CapsLock, or we toggled it remotely due to it being out of sync
+        // we clear the remote state so that we don't send duplicate or spurious fixes,
+        // since it may take some time to receive the new remote CapsLock state.
+        if (code == 'CapsLock') {
+            this._remoteCapsLock = null;
+        }
+        if (this._remoteCapsLock !== null && this._remoteCapsLock !== e.getModifierState('CapsLock')) {
+            Log.Warn("Fixing remote caps lock");
+
+            this._sendKeyEvent(KeyTable.XK_Caps_Lock, 'CapsLock', true);
+            this._sendKeyEvent(KeyTable.XK_Caps_Lock, 'CapsLock', false);
+            // We clear the remote capsLock state when we do this to prevent issues with doing this twice
+            // before we receive an update of the the remote state.
+            this._remoteCapsLock = null;
+        }
+
+        // Same goes for numlock is almost the same, except getModifierState for NumLock doesn't work on Mac,
+        // since Macs don't have a numlock key. So we just never try to fix numlock state on Mac.
+        // Also isn't supported for iOS.
+        if (code == 'NumLock') {
+            this._remoteNumLock = null;
+        }
+        if (!browser.isMac() && !browser.isIOS() && this._remoteNumLock !== null && this._remoteNumLock !== e.getModifierState('NumLock')) {
```

Comment:
It would be nice to avoid platform checks. I don't suppose macOS/iOS set anything on the event that makes it possible to detect that NumLock isn't supported?

Also, if we want to mimic current QEMU behaviour, shouldn't we try to always force NumLock on when on a macOS system?

Reply:
I haven't found a nicer way to detect lack of support no. Unfortunately passing an unsupported value to getModifierState returns false, exactly like numlock being off.

I could change it so numlock locally always counts as on instead for mac/ios yes. The reason I didn't, is that I don't think that behavior is better. Our getKeysym translates 'Clear' to 'NumLock' so that we specifically can toggle remote NumLock even on a mac. I think being able to toggle remote numlock, and use both states is better than not being able to, if the client is a mac, even if you can't see the state on your local keyboard. I am willing to change it though, if you disagree.